File: Stage31/Python/Problem7579.py
import sys

input = sys.stdin.readline

def main():
    N, M = map(int, input().split())
    memory = list(map(int, input().split()))
    cost = list(map(int, input().split()))

    max_cost = sum(cost)
    dp = [0] * (max_cost + 1)

    for i in range(N):
        for c in range(max_cost, cost[i] - 1, -1):
            dp[c] = max(dp[c], dp[c - cost[i]] + memory[i])

    for c in range(max_cost + 1):
        if dp[c] >= M:
            print(c)
            return

if __name__ == "__main__":
    main()

File: Stage31/C/Problem7579.c
#include <stdio.h>

int memory[101];
int cost[101];
int dp[10001];

int main(void) {
    int N, M;
    scanf("%d %d", &N, &M);

    for (int i = 0; i < N; i++) {
        scanf("%d", &memory[i]);
    }

    int maxCost = 0;
    for (int i = 0; i < N; i++) {
        scanf("%d", &cost[i]);
        maxCost += cost[i];
    }

    for (int i = 0; i < N; i++) {
        for (int c = maxCost; c >= cost[i]; c--) {
            if (dp[c] < dp[c - cost[i]] + memory[i]) {
                dp[c] = dp[c - cost[i]] + memory[i];
            }
        }
    }

    for (int c = 0; c <= maxCost; c++) {
        if (dp[c] >= M) {
            printf("%d\n", c);
            break;
        }
    }

    return 0;
}